scoring/score.py

```diff
@@ -1,14 +1,66 @@
+import collections
+
+
 class InvalidScoresheetException(Exception):
     pass
 
 
+TOKEN_HOME_CORNERS = 'PGYO'
+EXPECTED_TOKEN_COUNTS = {x: 5 for x in TOKEN_HOME_CORNERS}
+
+ZONES = list(range(len(TOKEN_HOME_CORNERS))) + ['other']
+
+
 class Scorer(object):
     def __init__(self, teams_data, arena_data):
         self._teams_data = teams_data
         self._arena_data = arena_data
 
+    def _score_team(self, team_info):
+        zone_id = team_info['zone']
+        own_token = TOKEN_HOME_CORNERS[zone_id]
+
+        token_string = self._arena_data[zone_id]['tokens'].replace(' ', '')
+
+        score = 1 if team_info.get('moved') else 0
+
+        for token in token_string:
+            if token == own_token:
+                score += 2
+            else:
+                score -= 1
+
+        return score
+
     def calculate_scores(self):
-        raise NotImplementedError()
+        return {
+            team_id: self._score_team(team_info)
+            for team_id, team_info in self._teams_data.items()
+        }
+
+    def validate(self, extra_data):
+        for team_id, team_info in self._teams_data.items():
+            if team_info.get('moved', False) and not team_info['present']:
```

Comment:
Okay, sounds perfectly acceptable to me :+1:

Reply:
Hrm, having just checked SRComp, it looks like it doesn't actually ensure that the `'present'` key is actually present. I think this is still ok though since our score records end up being automatically generated (by [srcomp-scorer](https://github.com/PeterJCLaw/srcomp-scorer)) and so will have the key.